modules/host/update.go

```diff
@@ -313,7 +295,25 @@ func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
 		h.log.Println(err)
 	}
 
-	// Handle the list of action items.
+	// Handle the list of action items. Action items require host locks and
+	// potentially require waiting for long periods of time while various
+	// network communications finish. To prevent the host lock from being held
+	// while the network I/O is blocking the action item execution, the action
+	// items are handled in their own thread. A wait group is used to prevent
+	// the host from releasing ThreadGroup control until all threads that this
+	// function have spawned have completed.
+	wg := new(sync.WaitGroup)
+	// The host should not release thread group control until all spawned
+	// threads have completed execution, though waiting for them should not
+	// block this function. The host should also not release thread group
+	// control until this function has terminated, hence the use of 'defer
+	// func() { go func() }' to coordinate waiting and done-ing.
+	defer func() {
+		go func() {
+			wg.Wait()
```

Comment:
I guess I only meant "unbounded" in the sense that you don't have any explicit control over how many goroutines will be running simultaneously; the only bound is the number of action items. In practice it will probably be a small number.
The reason it's weird is that the `tg.Add` and `tg.Done` are happening in separate threads. That's definitely convention-breaking. But I don't see an easy way around it if you're explicitly coding with `Flush()` in mind.
For example, this feels much clearer to me:
```go
func ProcessConsensusChange() {
	tg.Add()
	defer tg.Done()

	// collect action items...

	// handling action items takes time; spawn a new goroutine so we don't block
	go h.threadedHandleActionItems(actionItems)
}

func (h *Host) threadedHandleActionItems(actionItems []types.FileContractID) {
	tg.Add()
	defer tg.Done()

	// handle action items in parallel
	var wg sync.WaitGroup
	wg.Add(len(actionItems))
	for _, item := range actionItems {
		go h.threadedHandleActionItem(item, &wg)
	}
	wg.Wait()
}
```
...but then you need some way of delaying the call to `Flush()` until the `Add` in `threadedHandleActionItems`  has been called.